pkg/kudoctl/kudoinit/setup/setup.go

```diff
@@ -88,13 +88,16 @@ func (i *Installer) PreUpgradeVerify(client *kube.Client, result *verifier.Resul
 		return nil
 	}
 
-	// Step 2 - Verify that any migration is possible
-	migrations := requiredMigrations()
+	// Step 2 - Verify that all migrations can run (with dryRun)
+	migrations := requiredMigrations(client, true)
 	clog.V(1).Printf("Verify that %d required migrations can be applied", len(migrations))
 	for _, m := range migrations {
-		if err := m.CanMigrate(client); err != nil {
+		if err := m.CanMigrate(); err != nil {
```

Comment:
This will fail when this is run on already migrated resources, right?
IIUC, we must first check what is being migrated. And the list of required migrations is dependent on this. E.g., we have two migrations, migration 1 converts A to B, migration 2 converts B to C. So, A->B->C if both migrations are applied. Because migration 2 might have been added in a later version of KUDO than migration 1, there have to handle the case that there's already B, and only migration 2 needs to be applied. In addition to being able to handle A.
Does that make sense or am I missing something here?

Reply:
Yeah, this is difficult, because we don't really restrict *what* can be migrated. In this instance we operator on `OperatorVersion` and `Instance` CRs, but other migrations might add/remove labels/annotations on all deployed resources (ConfigMaps, Deployments, Statefulsets, etc..) or modify the KUDO prerequisites..

I agree that the A -> B -> C scenario is tricky. Generally, migrations should be idempotent, so if M1 migrates A -> B and M2 B -> C it should be no problem to repeatedly call M1, M2, M1, M1, M2 in any order. As soon as M1 and afterwards M2 is executed, there shouldn't be any more changes.

The only issue I might see is the "dry-run" and "canMigrate" cases. The "CanMigrate" func should really only check if there's something that prevents a migration from running - for example, a migration might want to update all CRs from `apiextensions/v1beta1` to `apiextensions/v1`, but the cluster actually doesn't support `apiextensions/v1` yet.

For "dry-run", i'm not sure we can fix this (without really really complex code). As dry-run doesn't execute the migration, it might happen that M1 dry-runs ok, and M2 doesn't find anything to migrate (no B exists yet). When the actual migration runs, M1 executes, converts A -> B, now M2 runs and finds a B, and fails.

Reply:
Thanks for mentioning idempotency! With "canMigrate" we can get this behavior. With that, calling M1(M1(A)) will work. Not sure if we want to return an error in case "canMigrate" fails though. A warning could be enough and also allow later migrations to be applied. So, M1(B) would create a warning, but still allow M2(B) which would create C. WDYT?

Reply:
I think that depends on the definition of "canMigrate". In my definition, `canMigrate` should only throw an error if it is *sure* that the migration cannot be executed. I'm reluctant to only use a warning in migrations, I think it should be "all or nothing".

But I assume this will evolve in the future with added migrations.

Reply:
Also - There is no `M1(B)` or `M2(B)`, the `canMigrate` is a global check on `M1` and `M2`, but it's not specific to a resource.